[AddressLowering] Handle indirectly yielded values used outside of coroutine range.

nate-chandler opened the pull request:
The `begin_apply` instruction introduces storage which is only valid until the coroutine is ended via `end_apply`/`abort_apply`.  Previously, AddressLowering assumed the memory was valid "as long as it needed to be": either as an argument to the function being lowered (so valid for the whole lifetime) or as a new `alloc_stack` (whose lifetime is determined by AddressLowering itself).  The result was that the storage made available by the `begin_apply` was used after the `end_apply`/`abort_apply` when there were uses of [a copy of] [a projection of] the yield after the `end_apply`/`abort_apply`.

Here, the behavior is fixed.  Now, the storage is used "as soon as possible".  There are two cases:

(1) If an indirect value's ownership is transferred into the caller (i.e. its convention is `@in` or `@in_constant`), the value is `copy_addr [take]`n into caller-local storage when lowering the `begin_apply`.

(2) If an indirect value's ownership is only lent to the caller (i.e. its convention is `@in_guaranteed`), no equivalent operation could be done: there's no `copy_addr [borrow]`; on the other hand, there's no need to do it--uses of the value must have occurred within the coroutine's range.  Instead, it's necessary to disable the store-copy optimization in this case: storage must be allocated for a copy of [a projection of] the yielded value.

nate-chandler committed:
[AddressLowering] Fix indirect coro yield handling

The begin_apply instruction introduces storage which is only valid until
the coroutine is ended via end_apply/abort_apply.  Previously,
AddressLowering assumed the memory was valid "as long as it needed to
be": either as an argument to the function being lowered (so valid for
the whole lifetime) or as a new alloc_stack (whose lifetime is
determined by AddressLowering itself).  The result was that the storage
made available by the begin_apply was used after the
end_apply/abort_apply when there were uses of [a copy of] [a projection
of] the yield after the end_apply/abort_apply.

Here, the behavior is fixed.  Now, the storage is used "as soon as
possible".  There are two cases:

(1) If an indirect value's ownership is transferred into the
caller (i.e. its convention is `@in` or `@in_constant`), the value is
`copy_addr [take]`n into caller-local storage when lowering the
begin_apply.

(2) If an indirect value's ownership is only lent to the caller (i.e.
its convention is `@in_guaranteed`), no equivalent operation could be
done: there's no `copy_addr [borrow]`; on the other hand, there's no
need to do it--uses of the value must have occurred within the
coroutine's range.  Instead, it's necessary to disable the store-copy
optimization in this case: storage must be allocated for a copy of [a
projection of] the yielded value.

nate-chandler committed:
[SIL] Verified yielded addresses.

For yielded, nontrivial addresses, verify based on convention:
- guaranteed -> must be initialized at end_apply/abort_apply
- owned -> must be deinitialized at end_apply/abort_apply
- inout -> must be initialized at end_apply/abort_apply

nate-chandler committed:
[SIL] Marked @in_constant as owned.

Matches the behavior of the memory lifetime verifier (and
OperandOwnership which was updated to match the memory lifetime verify
at 65fe9101300e40b3c0d36773af4fb2ecd725b912).

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test source compatibility

atrick commented on lib/SIL/Utils/MemoryLocations.cpp:
```diff
@@ -173,6 +173,21 @@ void MemoryLocations::analyzeLocations(SILFunction *function) {
+          auto fnType = BAI->getSubstCalleeType();
+          SILArgumentConvention argConv(
+              fnType->getYields()[index].getConvention());
+          if (argConv.isIndirectConvention()) {
+            // These are not SIL addresses under -enable-sil-opaque-values
+            if (!argConv.isInoutConvention() &&
+                !function->getConventions().useLoweredAddresses())
```
You should never need to check `useLoweredAddresses()` when querying the argument convention. For code like this that cares about the SIL convention, you're supposed to use the SIL-level convention API, like `isSILIndirect`

nate-chandler replied:
Fixed.

atrick commented:
What is `Indirect_In_Constant` equivalent to? `Indirect_In_Guaranteed` or `Indirect_In`? Our code-base seems to contradict itself. Can we just delete this convention in favor of one or the other?

atrick commented:
@aschwaighofer do you have any idea what this convention means: `ParameterConvention::Indirect_In_Constant`? If not, I suggest removing it immediately. It's own documentation contradicts its usage, and it needs to be checked everywhere we handle conventions, so keeping it in the code results in a continuing proliferation of bugs.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler committed:
[MemAccessUtils] Look thru nested for guar roots.

Previously, findGuaranteedReferenceRoots always stopped searching when
finding a begin_borrow, because it's not an ownership-forwarding
instruction.  Here, it is conditionally allowed to keep search through
the borrowee of that begin_borrow if it itself is guaranteed.

nate-chandler committed:
[AddressLowering] Soften assertion for end_borrows

When finding the PrunedLiveness on whose boundary end_borrows are to be
inserted, allow lifetime ending uses if they are end_borrows.  Such
end_borrows appear validly when an inner use is a nested begin_borrow.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test source compatibility

atrick commented:
LGTM
